### PR TITLE
feat: add reload listener for apm tracing config

### LINKED ISSUE #11381: Consume APM configuration over the elastic-agent control protocol
In order to enable APM tracing when APM Server is managed by Elastic Agent, APM Server needs to apply APM configuration settings when they're passed over the control protocol. This is necessary to enable tracing of APM Server in ESS and ECE.

These settings should be used to enable APM Server's self tracing functionality, and send the traces to the configured APM Server passed via the control protocol This may require that we change how the `apm.Tracer` object is created and passed to internal objects as the config may change during runtime.

References:
- elastic-agent-client changes: https://github.com/elastic/elastic-agent-client/pull/82
- https://github.com/elastic/elastic-agent/issues/2612

### PR DESCRIPTION
**Pull request by kyungeunni**
## Motivation/summary
~depending on https://github.com/elastic/beats/pull/40030~

Subscribe to the APMConfig changes from the reloader and hot-reload the runner with the updated tracing configuration.
This will enable self-instrumentation for APM Server in managed mode.

## Checklist

- [x] Update [CHANGELOG.asciidoc](https://github.com/elastic/apm-server/blob/main/CHANGELOG.asciidoc)
- [ ] Documentation has been updated
- [ ] Test with elastic-agent with the configuration written in `elastic-agent.yml` => will be done in separately

For functional changes, consider:
- Is it observable through the addition of either **logging** or **metrics**?
- Is its use being published in **telemetry** to enable product improvement?
- Have system tests been added to avoid regression?

## How to test these changes
Follow instructions in `Building an Elastic Agent container image with a locally built APM Server`  and `Running an Elastic Agent container with a locally built APM Server` section in `TESTING.md`, update the `elastic-agnet.yml` to see the config is received and it's passed down to tracing correctly.

## Related issues
- #11381
- https://github.com/elastic/beats/pull/40030

### CODE REVIEW

File: internal/beatcmd/reloader_test.go [hunk 1]
```diff
@@ -104,6 +104,10 @@ func TestReloader(t *testing.T) {
 	assert.NoError(t, err)
 	assertNoReload() // an output must be set
 
+	err = reload.RegisterV2.GetReloadableAPM().Reload(nil)
```
Comment (kyungeunni) [inline internal/beatcmd/reloader_test.go:107]: Reload can pass `nil`: https://github.com/elastic/beats/pull/40030/files#diff-dd024b66f2ed585cbbfce1d8a0aeba55183ab0d04bed2a67f7738f62b5658e63R899

**Review by axw — APPROVED** [review]
LGTM. Let's follow up with some manual testing before closing the issue please.

There's a couple of fields in APMConfig which I don't think we'll honour (`tls`, `global_labels`), but those should probably be handled in libbeat: https://github.com/elastic/elastic-agent-client/blob/dd646033eae6ef77cf6ff5d297e0d5513d08ac7f/pkg/proto/elastic-agent-client.pb.go#L1200-L1205

**Comment by kyungeunni** [comment]
> I don't think we'll honour (tls, global_labels), but those should probably be handled in libbeat

Yes, `instrumentation` in beats need to be updated to accept those fields. I made some changes in beats repository accepting those so I will make a PR for it as a follow up.

**Comment by kyungeunni** [comment]
## Follow up: Manual testing
I've done the manual testing by running a locally built elastic-agent image with the APM server binary that contains https://github.com/elastic/apm-server/pull/13514 which subscribes to the APMConfig reload and applies the given APMConfig for the self-instrumentation. 

### Setup
1. Run Elasticsearch and Kibana locally by running `docker-compose up -d`
1. Build the Elastic Agent image with the local APM Server:
```
Keep the image's name and tag
referring it as `$LOCAL_IMAGE` in the later example.(looks like `elastic-agent-systemtest:${snapshotversion}`)

$ cd systemtest/cmd/buildapm
$ go run main.go -arch arm64
```
1. Go to local Kibana(http://localhost:5601/app/fleet/agents), create agent policy  by clicking Add Agent button. Grab Enrollment token and policy ID from the snippet to install Elastic Agent
1. Create another docker-compose.yml file in a temp dir to run the Elastic Agent with the image built from above:
```
touch ~/temp/docker-compose.yaml
```
paste below, with replaced value for highlighted texts
```yaml
version: '3.9'
services:
  elastic-agent:
    image: **$LOCAL_IMAGE**
    networks:
      - network1
    ports:
      - 8220:8220
      - 8200:8200
    user: root
    environment:
      FLEET_SERVER_ENABLE: 'true'
      FLEET_SERVER_ELASTICSEARCH_HOST: http://elasticsearch:9200
      FLEET_SERVER_SERVICE_TOKEN: **Enrollment token**
      FLEET_SERVER_POLICY_ID: **fleet-server-policy**
networks:
  network1:
    name: apm-server_default
    external: true
```
1. run `docker-compose up -d`
1. see Elastic Agent is running in Health state, with the confirmation from Fleet UI in Kiabana. Once done, enable the APM Integration in Kibana UI. Confirm the logs to see APM Server is running successfully.

### Check for the Self Instrumentation
I recommend using docker-desktop, and I'm sharing this instructions based on the session I've done with docker-desktop.
1. In the logs, search for keyword `APM Instrumentation`, the latest entry should show `APM Instrumentation disabled`. The instrumentation is disabled by default.
1. Update the local configuration. APM Config can only be modified by updating `elastic-agent.yml` and the container needs to be restarted. 
update `/usr/shared/elastic-agent/state/elastic-agent.yml` to enable the tracing:
```
agent.monitoring:
  # enabled turns on monitoring of running processes
  enabled: true
  traces: true
  apm:
    environment: "kyungeun"
```
should look like this: 
image

> NOTE:
> Mounting modified elastic-agent.yml to `/usr/shared/elastic-agent/elastic-agent.yml` isn't honored, the configs are ignored and a new config is created in `/usr/shared/elastic-agent/state/elastic-agent.yml`. seems to be happening when in managed mode.


1. save the change and restart the container to take this change in place.
1. In the logs, search for the keyword `APM Instrumentation`, the latest entry should show `APM Instrumentation is enabled`. and you should see the request logs sent internally (source IP 127.0.0.1). You can also see the reloader gets the configs correctly in the logs.

**Pull request merged**